src/main/java/BallVision.java

```diff
@@ -10,107 +11,234 @@ Basic vision algorithm for detecting balls (powercells)
+            this.mat = mat;
+            this.contours = contours;
+        }
+    }
+
+    private static TargetLocation target;
+
+    //Load test images, process, and write output images. To-Do: Replace with camera input
     public static void main(String[] args) {
         System.loadLibrary(Core.NATIVE_LIBRARY_NAME);
-        for (int i = 1; i<10; i++){
-            String image= "src/main/java/BallImage";
-            image = image + i + ".jpg";
-            Mat input = Imgcodecs.imread(image);
-            Mat output = algorithm(input);
-            String output_fname = "Output Image " + i + ".jpg";
-            Imgcodecs.imwrite(output_fname, output);
+
+        if (args[0] == "TEST") {
+            long hm_images = 18;
+
+            // test using test_inputs, writing to test_outputs
+            for (int i = 1; i <= hm_images; i++) {
+                Mat input = Imgcodecs.imread("src/test_inputs/BallImage" + i + ".jpg");
+                Mat output = decoratePipelineOutput(ballDetectPipeline(input));
+                Imgcodecs.imwrite("src/test_outputs/Output Image " + i + ".jpg",
+                        output);
+            }
+            System.out.println("Done");
+        } else {
+            Mat image = Imgcodecs.imread(args[0]);
+            MatContourPair result = ballDetectPipeline(image);
+
+            double yaw = getTargetYaw(
+                    findBestBallTarget(result.contours),
+                    new Size(result.mat.width(), result.mat.height()),
+                    // the following are copied from HighGoalVision.java
+                    Math.toRadians(68.5),
+                    16.0,
+                    9.0
+            );
         }
-        System.out.println("Done");
+    }
+
+    public static TargetLocation process(Mat mat) {
+        MatContourPair matContourPair = ballDetectPipeline(mat);
+        Rect best = findBestBallTarget(matContourPair.contours);
+
+        double yaw = getTargetYaw(
+                best,
+                new Size(matContourPair.mat.width(), matContourPair.mat.height()),
+                Math.toRadians(68.5),
+                16.0,
+                9.0
+        );
+
+        return new TargetLocation(
+                best.x,
+                best.y,
+                best.width,
+                best.height,
+                yaw,
+                0 // pitch is irrelevant so just pass 0
+        );
     }
 
     /* calculate yaw or pitch (in radians) */
     static double calcAngle(long val, long centerVal, double focalLen) {
-        return Math.atan(((double)(val - centerVal)) / focalLen);
+        return Math.atan(((double) (val - centerVal)) / focalLen);
     }
 
-    public static Mat algorithm(Mat input) {
+    public static MatContourPair ballDetectPipeline(Mat input) {
         // Resize
         Mat resizeInput = input;
         double resizeWidth = 320.0;
-        double resizeHeight = 240.0;
+        double resizeHeight = 181.3;
```

Comment:
Why are you resizing to a non integer image size? (I think opencv probably just rounds anyway).

Note: you don't need to worry about matching the camera's aspect ratio in your sizing choice (if that is where 181.3 came from), as long as you pass 16/9 to `getTargetYaw` (which it looks like you did).

Reply:
In all honesty I do not remember where those values came from because I wrote them in early 2020.

Reply:
I would use 320.0 x 240.0 (standard + small enough to be quick to process) unless you have a reason not too.

You may have to adjust the height threshold of 100 (`ballHeightForAutoIntake` in 2020-robot-code) if you change this.